File: binarySearch/rotated/searchInRotatedSortedArr2.java
public int search(int[] A, int target) {
        if (A == null || A.length == 0) {
            return -1;
        }

        int start = 0;
        int end = A.length - 1;
        int mid;

        while (start + 1 < end) {
            mid = start + (end - start) / 2;
            if (A[mid] == target) {
                return mid;
            }
            //for duplicate element
            if(A[mid] == A[start]){
              start++;
            }else if (A[start] < A[mid]) { // situation 1: left is sorted
                if (A[start] <= target && target <= A[mid]) {
                    end = mid;
                } else {
                    start = mid;
                }
            // situation 2: right is sorted A[start] > A[mid]
            } else {
                if (A[mid] <= target && target <= A[end]) {
                    start = mid;
                } else {
                    end = mid;
                }
            }
        } // while
        if (A[start] == target) {
            return start;
        }
        if (A[end] == target) {
            return end;
        }
        return -1;
    }

File: binarySearch/rotated/searchInRotatedSortedArr2.py
def searchRotatedSortedArr2(arr, target):
    # with duplicate
    if arr is None or not arr:
        return -1

    left, right = 0, len(arr)-1
    while left + 1 < right:
        mid = left + (right- left)//2

        if arr[mid]== target:
            return mid
        elif arr[left] == arr[mid]:
            left += 1   # wrong: left = mid   13111 target 3
        elif arr[left] < arr[mid]:
            if arr[left] <= target and target <= arr[mid]:
                right = mid
            else:
                left = mid
        else:
            if arr[mid] <= target and target <= arr[right]:
                left = mid
            else:
                right = mid

    if arr[left] == target or arr[right] == target:
        return left
    return -1
